### ISSUE TITLE
Help with extension of Gedmo mapping parsers

### ISSUE THREAD

**Issue by ludekvodicka** [Atlantic18/DoctrineExtensions#1428]
Hi,

We're looking for programmer(s) who will help us extend Gedmo schema definition mapping parsers (for xml, yml, annotations).

We would like to implement a second version of schema format which will be similar to original Doctrine1 actAs behaviors. The reason for this is that we would like to add Gedmo support to our tool - Skipper (http://www.skipper18.com), which helps developers to design their ORM schema via visual editor.

We have received several requests to implement support of Gedmo extensions in our application, but unfortunately it is not possible to create it for current definition format because of its scattered nature. 

We already discussed the implementation with original DoctrineExtensions author Gediminas but unfortunately he didn't have a spare time for this. Because of that we would like to ask here and address all maintainers and contributors.

In exchange for that we're offering free licenses of our tool for the programmer or his team as well as some reasonable cash reward. We are interested in long-term cooperation, which we believe will help existing Gedmo users to use our application more smoothly as well as to introduce Gedmo extension to all our existing users. 

Below you can find our proposed technical specification. 
## Technical specification:

We're aware that it's not possible to change current version of schema format so we're proposing some kind of parallel version 2.0 format. Such format has to support all current features and should be covered by unit tests. Because an example is worth a thousand words, here is in a short form what would we like to achieve:

For example current form of Blameable format is following:

```
/**
 * @ODM\Document(collection="articles")
 */
class Article
{
   /**
     * @ORM\Column(type="string", length=128)
     */
    private $title;

    /**
     * @ORM\Column(name="body", type="string")
     */
    private $body;

    /**
     * @var string $createdBy
     *
     * @ODM\String
     * @Gedmo\Blameable(on="create")
     */
    private $createdBy;

    /**
     * @var string $updatedBy
     *
     * @ODM\String
     * @Gedmo\Blameable
     */
    private $updatedBy;

   /**
     * @var string $contentChangedBy
     *
     * @ORM\Column(name="content_changed_by", type="string", nullable=true)
     * @Gedmo\Blameable(on="change", field={"title", "body"})
     */
    private $contentChangedBy;
}
```

and we would like to have for example following format:

```
/** 
 * @ODM\Document(collection="articles")
 * @Gedmo\Blameable(
 *          createdBy={on="create"}, 
 *          updatedBy, 
 *          contentChangedBy={on="change",field={"title","body"}
 *    )
*/
```

This is only the example of possible format, we're open to discussion. The final solution of new schema definition isn't currently given and will be adjusted during our cooperation. The only requirement is to have whole behavior defined in one place in header of entity instead of several places like in existing format.

Mapping drivers (parsers) for each format are located in Gedmo/ExtensionName/Mapping/Driver (https://github.com/Atlantic18/DoctrineExtensions/blob/master/lib/Gedmo/Blameable/Mapping/Driver/Yaml.php). We're proposing to extend these parser classes to support both versions of format so it will be easy for future extension to update both formats.

Based on current analysis it shall not be hard to implement this feature. But we believe it's better to let implement and maintain this extension to someone who knows Gedmo extension better than implement it by ourselves.

In case of any questions please ask me here or send me direct email to ludek.vodicka@skipper18.com

Thank you for your time
Ludek
Chief Skipper developer

**Comment by Ocramius** [comment]
It seems to me like there is some issue with how documents are mapped in skipper, rather than with gedmo itself. Why do you need to have extension definitions at class-level definition?

Besides that, could you make an example with XML and what issues you'd have there?

**Comment by ludekvodicka** [comment]
Current version of Gedmo mapping isn't definitely wrong or bad. Why we're trying to extend it is that the current format is very hardly maintainable through Skipper. 

Originally, the architecture of our application was designed for Doctrine1 definitions and as time went we continuously extended our application, added Doctrine2 support, annotations support etc. But unfortunately current Gedmo extension is completely different from original Doctrine "actAs" concept and because of that it's difficult for us to add Gedmo support to our extension.

To provide you with an example in XML form, this is current form of Gedmo definitions:

    <entity name="Mapping\Fixture\Xml\Blameable" table="blameables">
            <id name="id" type="integer" column="id">
                <generator strategy="AUTO"/>
            </id>
    
            <field name="createdBy" type="string">
                <gedmo:blameable on="create"/>
            </field>
            <field name="updatedBy" type="string">
                <gedmo:blameable on="update"/>
            </field>
            <field name="publishedBy" type="string" nullable="true">
                <gedmo:blameable on="change" field="status.title" value="Published"/>
            </field>
    
            <many-to-one field="status" target-entity="Status">
                <join-column name="status_id" referenced-column-name="id"/>
            </many-to-one>
        </entity>

But what we would need/appreciated is something like this (final format could be different, the only requirement is "centralized" actAs element):

    <entity name="Mapping\Fixture\Xml\Blameable" table="blameables">
            <actAs>
                <blameable>
                   <fields>
                      <field on="create">createdBy</field>
                      <field on="update">updatedBy</field>
                      <field on="change"  field="status.title" value="Published">publishedBy</field>
                 </fields>
               </blameable>
            </actAs>
         
            <id name="id" type="integer" column="id">
                <generator strategy="AUTO"/>
            </id>
    
            <field name="createdBy" type="string"/>
            <field name="updatedBy" type="string"/>
            <field name="publishedBy" type="string" nullable="true"/>
    
            <many-to-one field="status" target-entity="Status">
                <join-column name="status_id" referenced-column-name="id"/>
            </many-to-one>
        </entity>


Such format would very simplify the import process because all behavior configurations would be available in single place together rather than distributed on several places. Implementation of current Gedmo format would mean to implement also an additional logic to correct updating all definitions on correct places. 

I hope it's more clear what we're trying to achieve with this change.

**Comment by Ocramius** [comment]
I understand where you are going, but:

 - changing Gedmo now is absolutely no-go, given the userbase and BC concerns
 - you seem to be fixing issues in an external component by just changing the core functionality instead
 - this sort of issue is handled by Gedmo as a "metadata loading pass"

Gedmo metadata is indeed completely separate metadata, and smashing it into the ORM metadata is probably a mistake in first place, but this issue seems to be tackling the problem from the wrong end.

So basically, `My.Foo.EntityName.dcm.xml` and `My.Foo.EntityName.gedmo.xml` would probably be the best approach to this (also from a DTD/XSD perspective).

Not sure about annotations, but moving everything to class-level there seems to be a huge mistake, and goes totally against how annotations are supposed to work (annotations are not "correct", they are more "friendly")

**Comment by ludekvodicka** [comment]
I definitely agree that it's not possible only to change the current behavior of Gedmo mapping without backward compatibility. What I'm trying to achieve is to prepare the second (parallel) version.

I think that solution you're mentioning is great (standalone .dcm.xml for original Doctrine definitions and standalone .gedmo.xml file for gedmo configurations). The same could be used for yml too.

But regarding the annotation, I think that both formats (current and proposed one) could coexist. As you wrote, there isn't single "correct" way. There is already a few similar class-level @ORM annotations (for example @Indexes or @AttributeOverrides) and it's definitely better than separated gedmo.php file.

**Comment by Ocramius** [comment]
> There is already a few similar class-level `@ORM` annotations (for example `@Indexes` or `@AttributeOverrides`) and it's definitely better than separated gedmo.php file.

Yeah, but those are indeed table-level additions, not column-level additions :-) There is some parallelism between the [target](https://github.com/doctrine/annotations/blob/f25c8aab83e0c3e976fd7d19875f198ccf2f7535/lib/Doctrine/Common/Annotations/Annotation/Target.php#L33-L37) of an annotation and what effect it has at db-level.
